File: mapM3uGuide.pl
#!/usr/bin/perl
use Data::Dumper;

$M3U_FILE = $ARGV[0] || die "specifiy m3u playlist";
$GUIDE_FILE = $ARGV[1] || die "specifiy xml guide file";

#load the contents of the files
my $m3uContent = do{local(@ARGV,$/)=$M3U_FILE;<>};
my $guideContent = do{local(@ARGV,$/)=$GUIDE_FILE;<>};

my %playlist; #map of playlist ID to playlist description/name
my %guide; #map of guide id to guide name
my %playlistToGuide; #mappig if playlist id to guide id, for the new guide


while ($m3uContent =~ /tvg-id\=\"(\w+.+?)\".+tvg-name=\"(\w+.+?)\"/g)
{
	my $m3uID=$1;
	my $m3uName=$2;
	$m3uName =~ s/\W+/ /g;
	push(@{$playlist{$m3uID}},lc($m3uName));
}
#print Dumper(%playlist);

while ($guideContent =~ /channel id=\"(\w+.+?)\"/g)
{
	my $guideID=$1;
	my $guideIDClean = $guideID;;
	$guideIDClean =~ s/\W+/ /g;
	$guide{$guideID} = lc $guideIDClean;
}


for my $pid (keys %playlist)
{
	for my $playlistName (@{$playlist{$pid}})
	{
		print STDERR "$pid<$playlistName>\n";
		@playlistTokens = split /\s+/,$playlistName;
		for my $gid (keys %guide)
		{
			my @guideTokens = split /\s+/,$guide{$gid};
			my $found = scalar @playlistTokens;
			for my $ptoken (@playlistTokens)
			{
				$found-- if $ptoken ~~ @guideTokens;		
			}

			#we have matched all tokens in the playlist name/desc
			if($found == 0)
			{
				print STDERR "\tMATCH: $pid<$playlistName}> = $gid<$guide{$gid}>\n";

				#we may have found a match that was better than the previous match, if its a closer match, replace it
				if($playlistToGuide{$pid})
				{
					my @prevGuideTokens = split /\s+/,$guide{$playlistToGuide{$pid}};
					if($#guideTokens < $#prevGuideTokens)
					{
						print STDERR  "\t\tBETTER MATCH, REPLACING $playlistToGuide{$pid} $guide{$playlistToGuide{$pid}}\n";
						$playlistToGuide{$pid} = $gid;
					}
				}

				#otherwise this is the first and only match so far
				else
				{
					$playlistToGuide{$pid} = $gid;
				}
			}
		}
	}	
}

#print the mapping to debug, and then display the new remapped guide
for my $pid (sort keys %playlistToGuide)
{ 
	$oldgid = $playlistToGuide{$pid};
	print STDERR "[$pid] -> [$oldgid]\n";
	$guideContent =~ s/\Q"$oldgid"/\"$pid\"/g;
}
print $guideContent;

File: mapM3uguide.py
#!/usr/bin/env python3
import sys
import re
import xml.etree.ElementTree



def readfile (fileName):
    text_file = open(fileName, "r")
    data = text_file.read()
    text_file.close()
    return data

def main() -> int:
    playlist = {}
    guide = {}
    playlistToGuide = {}
    m3uContent = readfile(sys.argv[1])
    guideContent = readfile(sys.argv[2])

    it = re.finditer("tvg-id\=\"(\w+.+?)\".+tvg-name=\"(\w+.+?)\"", m3uContent)
    for m in it:
        playlist.setdefault(m.group(1),[]).append(re.sub("\W+"," ", m.group(2)).lower())

    it = re.finditer("channel id=\"(\w+.+?)\"", guideContent)
    for m in it:
        guide[m.group(1)] = re.sub("\W+"," ", m.group(1).lower())

    for pid in playlist:
        for playlistName in playlist[pid]:
            playlistTokens = playlistName.split()
            for gid in guide:
                guideTokens = guide[gid].split()
                found = len(playlistTokens)
                for pToken in playlistTokens:
                    if pToken in guideTokens:
                        found -= 1

                if found == 0:
                    if pid in playlistToGuide:
                        if len(guideTokens) < len(guide[playlistToGuide[pid]].split()):
                            playlistToGuide[pid] = gid
                            print(pid,playlistTokens,"\t",gid,guideTokens,"BETTER MATCH",file = sys.stderr)
                    else:
                        playlistToGuide[pid] = gid 
                        print(pid,playlistTokens,"\t",gid,guideTokens,file = sys.stderr)

    #invert the dictionary for 2way lookup
    invert_playlistToGuide = {v: k for k, v in playlistToGuide.items()}

    for k in sorted(playlistToGuide):
        print(k,"->",playlistToGuide[k], file=sys.stderr)
    
    
    print("<?xml version=\"1.0\" encoding=\"UTF-8\" ?>")
    print("<tv generator-info-name=\"none\" generator-info-url=\"none\">")

    #walk the guide and replace ids. throw out things that are unmapped
    for event, element in xml.etree.ElementTree.iterparse(sys.argv[2], events=("start","end")):
        if element.tag == "programme" and event == "start":
            if element.get("channel") in invert_playlistToGuide:
                print(xml.etree.ElementTree.tostring(element).decode().replace(element.get("channel"),invert_playlistToGuide[element.get("channel")]))

        elif  element.tag == "channel" and event == "start":
           if element.get("id") in invert_playlistToGuide:
                print(xml.etree.ElementTree.tostring(element).decode().replace(element.get("id"),invert_playlistToGuide[element.get("id")])) 
        element.clear()

    print("</tv>")

    return 0

if __name__ == '__main__':
    sys.exit(main()) 


